README.md
V3toys yii2 api
===================================

http://www.v3toys.ru/index.php?nid=api

Installation
------------

The preferred way to install this extension is through [composer](http://getcomposer.org/download/).

Either run

```
php composer.phar require --prefer-dist v3toys/yii2-api "*"
```

or add

```
"v3toys/yii2-api": "*"
```

How to use
----------

```php
//App config
[
    'components'    =>
    [
    //....
        'v3toys' =>
        [
            'class'             => '\v3toys\yii2\api\Api',
            'url'               => 'http://www.v3toys.ru/pear-www/Kiwi_Shop/api.php',
            'affiliate_key'     => 'fff',
            'timeout'           => 12,
        ],
    //....
    ]
]

```

Examples
----------

```php

$response = \Yii::$app->v3toys->send('getProductsDataByIds', [
    'products_ids' => 217070
]);

if ($response->isError)
{
    echo $response->error_code;
    echo $response->error_message;
}

if ($response->isOk)
{
    print_r($response->data);
}

```
___

> [![skeeks!](https://gravatar.com/userimage/74431132/13d04d83218593564422770b616e5622.jpg)](http://skeeks.com)  
<i>SkeekS CMS (Yii2) — fast, simple, effective!</i>  
[skeeks.com](http://skeeks.com) | [cms.skeeks.com](http://cms.skeeks.com) | [marketplace.cms.skeeks.com](http://marketplace.cms.skeeks.com)



CHANGELOG.md
CHANGELOG
==============

1.0.0-beta
-----------------
  * Добавлен заказ в 1 клик

1.0.0-alpha
-----------------
  * Can be used


ApiBase.php
<?php
/**
 * @author Semenov Alexander <semenov@skeeks.com>
 * @link http://skeeks.com/
 * @copyright 2010 SkeekS (СкикС)
 * @date 15.07.2016
 */
namespace v3toys\yii2\api;

use skeeks\cms\helpers\StringHelper;
use v3toys\yii2\api\models\ApiResponseError;
use v3toys\yii2\api\models\ApiResponseOk;
use yii\base\Component;
use yii\helpers\Json;
use yii\httpclient\Client;

/**
 * Class ApiBase
 * @property string $version read-only
 *
 * @see http://www.v3toys.ru/index.php?nid=api
 * @package v3toys\yii2\api
 */
abstract class ApiBase extends Component
{
    /**
     * версия api, на настоящий момент 0.4
     */
    const VERSION = '0.4';

    /**
     * @var string
     */
    public $url = 'http://www.v3toys.ru/pear-www/Kiwi_Shop/api.php';

    /**
     * @var null|string ключ аффилиата в системе V3Project, если он передается то контроль доступа происходит по IP+"affiliate_key"
     */
    public $affiliate_key = '';




    /**
     * @var int set timeout to 15 seconds for the case server is not responding
     */
    public $timeout = 30;

    /**
     * @param array $data
     * @param string $method
     *
     * @return ApiResponseOk|ApiResponseError
     */
    private function _send(array $data, $method = 'post')
    {
        $client = new Client([
            'requestConfig' => [
                'format' => Client::FORMAT_JSON
            ]
        ]);

        $response = $client->createRequest()
                ->setMethod($method)
                ->setUrl($this->url)
                ->addHeaders(['Content-type' => 'application/json'])
                ->addHeaders(['user-agent' => 'JSON-RPC PHP Client'])
                ->setData($data)
                ->setOptions([
                    'timeout' => $this->timeout
                ])
            ->send();
        ;

        //Нам сказали это всегда json. А... нет, все мы люди, бывает и не json )
        try
        {
            $dataResponse = (array) Json::decode($response->content);
        } catch (\Exception $e)
        {
            \Yii::warning("Json api response error: " . $e->getMessage() . ". Response: \n{$response->content}", self::className());
            //Лайф хак, вдруг разработчики апи оставили var dump
            if ($pos = strpos($response->content, "{"))
            {
                $content = StringHelper::substr($response->content, $pos, StringHelper::strlen($response->content));

                try
                {
                    $dataResponse = (array) Json::decode($content);
                } catch (\Exception $e)
                {
                    \Yii::warning("Api response error: " . $response->content, self::className());
                }
            }
        }


        if (!$response->isOk)
        {
            \Yii::error($response->content, self::className());
            $responseObject = new ApiResponseError($dataResponse);
        } else
        {
            $responseObject = new ApiResponseOk($dataResponse);
        }

        $responseObject->statusCode = $response->statusCode;

        return $responseObject;
    }

    /**
     * @param $method           вызываемый метод, список приведен далее
     * @param array $params     параметры соответствующие методу запроса
     *
     * @return ApiResponseError|ApiResponseOk
     */
    public function send($method, array $params = [])
    {
        $request = [
            'v'                 => static::VERSION,
            'affiliate_key'     => $this->affiliate_key,
            'method'            => $method,
            "params"            => $params
        ];

        return $this->_send($request);
    }

    /**
     * @return string
     */
    public function getVersion()
    {
        return static::VERSION;
    }
}

models/ApiResponse.php
<?php
/**
 * @author Semenov Alexander <semenov@skeeks.com>
 * @link http://skeeks.com/
 * @copyright 2010 SkeekS (СкикС)
 * @date 15.07.2016
 */
namespace v3toys\yii2\api\models;

use yii\base\Component;

/**
 * Описание общих полей запросов
 *
 * @property bool isError
 * @property bool isOk
 *
 * @see http://www.v3toys.ru/index.php?nid=api #1
 * @package v3toys\yii2\api\models
 */
abstract class ApiResponse extends Component
{
    /**
     * версия api, на настоящий момент 0.4
     * @var string
     */
    public $v;

    /**
     * ключ аффилиата в системе V3Project
     * @var string
     */
    public $affiliate_key;

    /**
     * вызыванный метод, список приведен далее
     * @var string
     */
    public $method;

    /**
     * данные соответствующие методу запроса
     * @var mixed
     */
    public $data;


    /**
     * Seerver response code
     * @var int
     */
    public $statusCode;

    /**
     * Ответны запрос ошибочный?
     * @return bool
     */
    abstract public function getIsError();

    /**
     * @return bool
     */
    public function getIsOk()
    {
        return !$this->isError;
    }
}

ApiV04.php
<?php
/**
 * @author Semenov Alexander <semenov@skeeks.com>
 * @link http://skeeks.com/
 * @copyright 2010 SkeekS (СкикС)
 * @date 15.07.2016
 */
namespace v3toys\yii2\api;
/**
 * Class ApiV04
 *
 * @package v3toys\yii2\api
 */
class ApiV04 extends ApiBase
{
    const VERSION = '0.4';

    /**
     * 3.1.5 Метод getStatus - возвращает все возможные статусы заказов в V3Project
     *
     * @return models\ApiResponseError|models\ApiResponseOk
     * "data": [
            {
                "id": 1,
                "title": "Принят"
            },
            {
                "id": 2,
                "title": "Отмена"
            }
        ]
     */
    public function getStatus()
    {
        return $this->send('getStatus');
    }


    /**
     *
     *  Метод getOrderStatus - возвращает статус заказа *** должен быть один из двух параметров (order_id или v3p_oid)
     *
     * "data": {
            "status_id": 599,
            "status": "Принят",
        }
     *
     * @param $v3toysOrderId
     * @return models\ApiResponseError|models\ApiResponseOk
     */
    public function getOrderStatusByV3toysId($v3toysOrderId)
    {
        return $this->send('getOrderStatus', [
            'v3p_oid' => $v3toysOrderId
        ]);
    }

    /**
     *  Метод getOrderStatus - возвращает статус заказа *** должен быть один из двух параметров (order_id или v3p_oid)
     *
     * "data": {
            "status_id": 599,
            "status": "Принят",
        }
     *
     *
     * @param $orderId
     * @return models\ApiResponseError|models\ApiResponseOk
     */
    public function getOrderStatusById($orderId)
    {
        return $this->send('getOrderStatus', [
            'order_id' => $orderId
        ]);
    }

    /**
     * 3.1.3 Метод getProductsDataByIds - получение данных о товаре по коду
     *
     * @param array $request
     * @return models\ApiResponseError|models\ApiResponseOk
     *
     * "data": [
            {
                "id": 109420,
                "deleted": 0,
                "quantity": 1,              //количеcтво
                "buy_price": 500,           //закупочная цена
                "base_price": 530,          //базовая цена
                "mr_price": 553,            //минимальная розничная рцена
                "price": 553,               //ориентировочная цена продажи
                "title": "Игровой набор..",
                "sku": "A8532/A8232",
                "barcode": "5010994804367",
                "brand": "Hasbro"
            }
        ]
     */
    public function getProductsDataByIds($request = [])
    {
        return $this->send('getProductsDataByIds', array_merge([
            'full'              => 1,       //признак получения полной информации (title, sky, barcode)
            'products_ids'      => 'all',   //Возможные значения: all: все товары >код: товары с кодом больше заданного код,код: товары заданных кодов код: товар заданного кода
            'limit'             => 100,     //количество получаемых товаров (по умолчанию 10000)
            'offset'            => 0,       //смещение при выборке товаров (по умолчанию 0)
            //'in_stock'        => 0,       //признак получения товаров в наличие:
            //'is_deleted'      => 0,       //признак получения удаленных товаров
        ], $request));
    }

    /**
     * {
            "order_id": "3423",
            "fake": 1,
            "created_at": "2014-02-03 15:34:23",
            "full_name": "Игорь",
            "phone": "79261234567",
            "email": "test@example.com",
            "products": [
                {
                    "product_id": 2352,
                    "price": 2246,
                    "quantity": 1
                },
                {
                    "product_id": 54352,
                    "price": 46,
                    "quantity": 2
                },
            ],
            "shipping_method": "PICKUP",
            "shipping_cost": 50,
            "shipping_data": {
                "city": "Орел",
                "point_id": 1
            },
            "comment": "Тестовый комментарий"
        }
     *
     * @param array $request
     *
     * @return models\ApiResponseError|models\ApiResponseOk
     */
    public function createOrder($request = [])
    {
        return $this->send('createOrder', $request);
        /*return $this->send('createOrder', array_merge([
            /*'full_name'             => "",
            'phone'             => "",
            'email'             => "",
            'products'             => [
                [
                    'product_id' => '',
                    'price' => '',
                    'quantity' => '',
                    'is_deleted' => '',
                ]
            ],  //имя клиента
            'shipping_method' => '',
            'shipping_cost' => '',
            'shipping_data' => [
                'city' => '',
                'address' => '',
            ],
        ], $request));*/
    }


    /**
     *
     * 3.2.1 Метод createMessage - создает заявку и возвращает ее номер
        ***Параметры этого метода полностью совпадают с данными ответа метода
     *
     * {
        "v": "0.4",
        "method": "getMessageDataById",
        "data": {
            "message_id": "3423",
            "fake": 1,
            "created_at": "2014-02-03 15:34:23",
            "full_name": "Игорь",
            "phone": "79261234567",
            "email": "",
            "products": [
                {
                    "product_id": 2352,
                    "price": 2246,
                    "quantity": 1
                },
            ],
            "comment": "Тестовый комментарий"
        }
    }
     *
     * @param array $request
     *
     * @return models\ApiResponseError|models\ApiResponseOk
     */
    public function createMessage($request = [])
    {
        return $this->send('createMessage', $request);
        /*return $this->send('createOrder', array_merge([
            /*'full_name'             => "",
            'phone'             => "",
            'email'             => "",
            'products'             => [
                [
                    'product_id' => '',
                    'price' => '',
                    'quantity' => '',
                    'is_deleted' => '',
                ]
            ],  //имя клиента
            'shipping_method' => '',
            'shipping_cost' => '',
            'shipping_data' => [
                'city' => '',
                'address' => '',
            ],
        ], $request));*/
    }


    /**
     * 3.2.2 Метод getMessageStatus - возвращает статус заявки
     * @param $messageId
     *
     * @return models\ApiResponseError|models\ApiResponseOk
     */
    public function getMessageStatus($messageId)
    {
        return $this->send('getMessageStatus', [
            'message_id' => $messageId
        ]);
    }
}